### PR TITLE
Write files in new directories

### PR DESCRIPTION
**Pull request by Brett208**
I think there is an issue with XFile::GetDirectory

It returns the filename without the extension in some cases instead of the directory.

### CODE REVIEW

File: test/Stream/FileWriter.test.cpp [hunk 1]
```diff
@@ -65,3 +65,13 @@ TEST(FileWriter, MoveConstructible) {
...
 }
+
+TEST(FileWriter, DirectoryDoesNotExist) {
+	EXPECT_NO_THROW(Stream::FileWriter writer("../NewDirectory/TestFile.temp"));
+	EXPECT_NO_THROW(Stream::FileWriter writer("./NewDirectory/TestFile.temp"));
+	EXPECT_NO_THROW(Stream::FileWriter writer("NewDirectory/TestFile.temp"));
+
+	EXPECT_NO_THROW(Stream::FileWriter writer("..\\NewDirectory\\TestFile.temp"));
+	EXPECT_NO_THROW(Stream::FileWriter writer(".\\NewDirectory\\TestFile.temp"));
+	EXPECT_NO_THROW(Stream::FileWriter writer("NewDirectory\\TestFile.temp"));
```
[the remark this answers is not in the document]
Comment (DanRStevens) [reply test/Stream/FileWriter.test.cpp]: Yes, on Linux, the Windows path separator leaves a file literally named "NewDirectory\TestFile.temp". The "\" is treated just like any other character as part of the filename. As such the tests pass, but aren't really doing what they were expected to do.

**Comment by DanRStevens** [comment]
The new tests should probably clean up the temporary files they are generating.

File: src/Stream/FileWriter.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ namespace Stream
 	{
 		// Create directory if it does not exist. ofstream will fail if directory does not exist.
 		auto directory = XFile::GetDirectory(filename);
-		if (!XFile::PathExists(directory)) {
+		if (directory != "" && !XFile::PathExists(directory)) {
```
[the remark this answers is not in the document]
Comment (Brett208) [reply src/Stream/FileWriter.cpp]: I also like that string::empty reads easier than doing a not equal comparison.

File: test/Stream/FileWriter.test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Brett208) [reply test/Stream/FileWriter.test.cpp]: I wanted to be thorough by testing both cases. However, we are just testing that a file is created in the new directory, so testing using both directory separators on Windows seems redundant.

**Comment by Brett208** [comment]
@DanRStevens would you be willing to help resolve the package install issues going on with the travis-ci tests?

Thanks,
Brett

File: src/Stream/FileWriter.cpp [hunk 3]
```diff
@@ -40,7 +40,12 @@ namespace Stream
 	{
 		// Create directory if it does not exist. ofstream will fail if directory does not exist.
 		auto directory = XFile::GetDirectory(filename);
-		if (!directory.empty() && !XFile::PathExists(directory)) {
+		if (!directory.empty() && !XFile::PathExists(directory)) 
+		{
+			if (openMode != OpenMode::CanOpenNew) {
```
Comment (DanRStevens) [inline src/Stream/FileWriter.cpp]: This should test for the specific flag with `&`, rather than the combination of all flags with exact equality/inequality.
```
# Short form
if (openMode & OpenMode::CanOpenNew) {
```
```
# Explicit form
if ((openMode & OpenMode::CanOpenNew) != 0) {
```
```
# Overly explicit form
if ((openMode & OpenMode::CanOpenNew) == OpenMode::CanOpenNew) {
```

I think I would prefer the first form.

Comment (DanRStevens) [reply src/Stream/FileWriter.cpp]: I think I might also suggest testing the `OpenMode` flag before checking for a non-empty directory component to the path, or checking the filesystem itself.

**Comment by DanRStevens** [comment]
I'm checking into the TravisCI stuff.

File: src/Stream/FileWriter.cpp (hunk 3 above)
Comment (Brett208) [reply src/Stream/FileWriter.cpp]: Thanks for catching my error on the flag check. Should be fixed with newest push.

If we check the OpenMode flag first, if a new directory is required, the code would bypass our custom error message and attempt to call ofstream. This would produce an error, but not as descriptive. Maybe I'm missing your point though...

-Brett

Comment (DanRStevens) [reply src/Stream/FileWriter.cpp]: Ahh, I didn't notice the custom error message.

**Comment by DanRStevens** [comment]
Should we re-purpose the `CanOpenNew` flag with this new meaning, or should we add a new flag to allow automatic directory creation? I don't mind using the `CanOpenNew` flag to guard this behaviour, as it would seem odd to create new directories if creating new files wasn't allowed. Though it seems conceivable that someone might want to allow creating new files, but not allow creating new directories.

I may be overthinking this though.

**Comment by Brett208** [comment]
I'm happy with the existing flag structure although I wouldn't oppose the new addition if you thought it was important.

**Comment by Brett208** [comment]
@DanRStevens are you okay if I merge this branch into master now.

Thanks,
Brett
